core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -484,13 +466,38 @@ class TransactionStateManager(brokerId: Int,
       responseCallback(responseError)
     }
 
-    replicaManager.appendRecords(
-      newMetadata.txnTimeoutMs.toLong,
-      TransactionLog.EnforcedRequiredAcks,
-      internalTopicsAllowed = true,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    getAndMaybeAddTransactionState(transactionalId) match {
+      case Left(err) =>
+        responseCallback(err)
+
+      case Right(None) =>
+        // the coordinator metadata has been removed, reply to client immediately with NOT_COORDINATOR
+        responseCallback(Errors.NOT_COORDINATOR)
+
+      case Right(Some(epochAndMetadata)) =>
+        val metadata = epochAndMetadata.transactionMetadata
+
+        metadata synchronized {
+          if (epochAndMetadata.coordinatorEpoch != coordinatorEpoch) {
+            // the coordinator epoch has changed, reply to client immediately with with NOT_COORDINATOR
+            responseCallback(Errors.NOT_COORDINATOR)
+          } else {
+            // do not need to check the metadata object itself since no current thread should be able to modify it
```

Comment:
About the locking, here is my reasoning: https://github.com/apache/kafka/pull/3082#discussion_r117591792